scripts/feature_tests/upgrade/combined_tests/1cp_1w_bootDiskImageANDK8sControllers_clusterLevel_upgrade.sh

```diff
@@ -0,0 +1,172 @@
+
+echo '' >~/.ssh/known_hosts
+
+start_logging "${1}"
+# Provision original nodes
+set_number_of_master_node_replicas 1
+set_number_of_worker_node_replicas 1
+
+provision_controlplane_node
+
+controlplane_is_provisioned
+controlplane_has_correct_replicas 1
+
+# apply CNI
+apply_cni
+
+provision_worker_node
+worker_has_correct_replicas 1
+# ----------- upgrade controlplne components ---------------
+
+cleanup_clusterctl_configuration
+
+buildClusterctl
+
+# Install initial version folder structure
+pushd /tmp/cluster-api-clone || exit
+cmd/clusterctl/hack/local-overrides.py
+popd || exit
+
+create_clusterctl_configuration
+
+# Create a new version
+createNextVersionControllers
+
+makeCrdChanges
+
+# show upgrade plan
+clusterctl upgrade plan
+# do upgrade
+clusterctl upgrade plan | grep "upgrade apply" | xargs | xargs clusterctl
+# shellcheck disable=SC2082
+if [ "${$?}" -ne '0' ]; then
+  clusterctl upgrade plan
+  echo "Upgrade of controllers with clusterctl has failed."
+  exit 1
+fi
+
+# Verify upgrade
+upgraded_controllers_count=$(kubectl api-resources | grep -Ec "kcp2020|ma2020")
+upgraded_bootstrap_crd_count=$(kubectl get crds \
+  kubeadmconfigs.bootstrap.cluster.x-k8s.io -o json | jq '.spec.names.singular' | wc -l)
+upgraded_capm3_controller_count=$(kubectl api-resources | grep -c m3c2020)
+
+if [ "${upgraded_controllers_count}" -ne 2 ]; then
+  log_error "Failed to upgrade cluster-api and controlplane components"
+  log_test_result "upgrade_CAPI_and_CAPM3_with_clusterctl.sh" "fail"
+  exit 1
+fi
+if [ "${upgraded_bootstrap_crd_count}" -ne 1 ]; then
+  log_error "Failed to upgrade control-plane-kubeadm components"
+  log_test_result "upgrade_CAPI_and_CAPM3_with_clusterctl.sh" "fail"
+  exit 1
+fi
+
+if [ "${upgraded_capm3_controller_count}" -ne 1 ]; then
+  log_error "Failed to upgrad infrastructure components"
+  log_test_result "upgrade_CAPI_and_CAPM3_with_clusterctl.sh" "fail"
+  exit 1
+fi
+
+sleep 30 # Wait for the controllers to be up and running
+
+health_controllers=$(kubectl get pods -A | grep -E "capm3-system|capi-kubeadm|metal3" | grep -vc 'Running')
+if [ "${health_controllers}" -ne 0 ]; then
+  log_error "Some of the upgraded controlplane componenets are not healthy"
+  log_test_result "upgrade_CAPI_and_CAPM3_with_clusterctl.sh" "fail"
+  exit 1
+fi
+
+# ----------- upgrade ironic image ------
+
+# Upgrade container images
+for container in ironic ironic-dnsmasq ironic-httpd mariadb; do
+  kubectl set image deployments metal3-ironic \
+    $container=quay.io/metal3-io/ironic:"${IRONIC_IMAGE_TAG}" -n metal3
+done
+kubectl set image deployments metal3-ironic \
+  ironic-inspector=quay.io/metal3-io/ironic-inspector:"${IRONIC_IMAGE_TAG}" -n metal3
+
+updated_ironic_image_count=$(kubectl get deployments metal3-ironic -n metal3 -o json |
+  jq '.spec.template.spec.containers[].image' | grep -c "${IRONIC_IMAGE_TAG}")
+
+if [ "${updated_ironic_image_count}" -lt 5 ]; then
```

Comment:
Should we leave hard coded count for ironic images or it is better to parameterize maybe?

Reply:
good comment.
I'll take a note and do that in a separate commit once upgrade tests are merged to metal3-dev-env.